packages/smart-contracts/src/contracts/BatchConversionPayments.sol

```diff
@@ -25,202 +24,173 @@ contract BatchConversionPayments is BatchNoConversionPayments {
-          metaConversionDetail.cryptoDetails.paymentReferences,
-          metaConversionDetail.cryptoDetails.feeAmounts,
-          _feeAddress
+      MetaDetail calldata metaDetail = metaDetails[i];
+      if (metaDetail.paymentNetworkId == 0) {
+        batchFeeAmountUSD += batchMultiERC20ConversionPayments(
+          metaDetail.requestDetails,
+          batchFeeAmountUSD,
+          pathsToUSD,
+          feeAddress
+        );
+      } else if (metaDetail.paymentNetworkId == 1) {
+        batchFeeAmountUSD += batchERC20Payments(
+          metaDetail.requestDetails,
+          pathsToUSD,
+          batchFeeAmountUSD,
+          feeAddress
         );
-      } else if (metaConversionDetail.paymentNetworkId == 2) {
-        batchMultiERC20Payments(
-          metaConversionDetail.cryptoDetails.tokenAddresses,
-          metaConversionDetail.cryptoDetails.recipients,
-          metaConversionDetail.cryptoDetails.amounts,
-          metaConversionDetail.cryptoDetails.paymentReferences,
-          metaConversionDetail.cryptoDetails.feeAmounts,
-          _feeAddress
+      } else if (metaDetail.paymentNetworkId == 2) {
+        batchFeeAmountUSD += batchMultiERC20Payments(
+          metaDetail.requestDetails,
+          pathsToUSD,
+          batchFeeAmountUSD,
+          feeAddress
         );
-      } else if (metaConversionDetail.paymentNetworkId == 3) {
+      } else if (metaDetail.paymentNetworkId == 3) {
         if (metaDetails[metaDetails.length - 1].paymentNetworkId == 4) {
           // Set to false only if batchEthConversionPayments is called after this function
           transferBackRemainingEth = false;
         }
-        batchEthPayments(
-          metaConversionDetail.cryptoDetails.recipients,
-          metaConversionDetail.cryptoDetails.amounts,
-          metaConversionDetail.cryptoDetails.paymentReferences,
-          metaConversionDetail.cryptoDetails.feeAmounts,
-          payable(_feeAddress)
+        batchFeeAmountUSD += batchEthPayments(
+          metaDetail.requestDetails,
+          pathsToUSD.length > 0,
+          batchFeeAmountUSD,
+          payable(feeAddress)
         );
         if (metaDetails[metaDetails.length - 1].paymentNetworkId == 4) {
           transferBackRemainingEth = true;
         }
-      } else if (metaConversionDetail.paymentNetworkId == 4) {
-        batchEthConversionPayments(metaConversionDetail.conversionDetails, payable(_feeAddress));
+      } else if (metaDetail.paymentNetworkId == 4) {
+        batchFeeAmountUSD += batchEthConversionPayments(
+          metaDetail.requestDetails,
+          pathsToUSD.length > 0,
+          batchFeeAmountUSD,
+          payable(feeAddress)
+        );
       } else {
-        revert('wrong paymentNetworkId');
+        revert('Wrong paymentNetworkId');
       }
     }
+    if (pathsToUSD.length > 0 && metaDetails.length > 1) {
+      // Set back to false, its default value
+      batchPaymentOrigin = false;
```

Comment:
I don't really understand the pattern for `batchPaymentOrigin`, can you clarify? Aren't there edgy situations where it remains true and it can be a problem?